apps/dashboard/src/components/workflow-editor/steps/configure-step-form.tsx

```diff
@@ -23,70 +24,137 @@ import { Input, InputField } from '@/components/primitives/input';
+  return merge(buildDefaultValuesOfDataSchema(step.controls.dataSchema ?? {}), step.controls.values);
+};
 
 type ConfigureStepFormProps = {
   workflow: WorkflowResponseDto;
   environment: IEnvironment;
   step: StepDataDto;
+  issues?: StepIssuesDto;
   update: (data: UpdateWorkflowDto) => void;
   updateStepCache: (step: Partial<StepDataDto>) => void;
 };
 
 export const ConfigureStepForm = (props: ConfigureStepFormProps) => {
-  const { step, workflow, update, updateStepCache, environment } = props;
-  const navigate = useNavigate();
-  const isCodeCreatedWorkflow = workflow.origin === WorkflowOriginEnum.EXTERNAL;
+  const { step, workflow, update, updateStepCache, environment, issues } = props;
+
+  const isSupportedStep = !UNSUPPORTED_STEP_TYPES.includes(step.type);
+  const isReadOnly = !isSupportedStep || workflow.origin === WorkflowOriginEnum.EXTERNAL;
+
+  const isTemplateConfigurableStep = isSupportedStep && TEMPLATE_CONFIGURABLE_STEP_TYPES.includes(step.type);
+  const isInlineConfigurableStep = isSupportedStep && INLINE_CONFIGURABLE_STEP_TYPES.includes(step.type);
 
+  const navigate = useNavigate();
   const [isDeleteModalOpen, setIsDeleteModalOpen] = useState(false);
 
   const onDeleteStep = () => {
     update({ ...workflow, steps: workflow.steps.filter((s) => s._id !== step._id) });
     navigate(buildRoute(ROUTES.EDIT_WORKFLOW, { environmentSlug: environment.slug!, workflowSlug: workflow.slug }));
   };
 
-  const defaultValues = {
-    name: step.name,
-    stepId: step.stepId,
-  };
+  const registerInlineControlValues = useMemo(() => {
+    return (step: StepDataDto) => {
+      if (isInlineConfigurableStep) {
+        return {
+          controlValues: calculateDefaultControlsValues(step),
+        };
+      }
+
+      return {};
+    };
+  }, [isInlineConfigurableStep]);
+
+  const controlsSchema = useMemo(
+    () => (isInlineConfigurableStep ? buildDynamicZodSchema(step.controls.dataSchema ?? {}) : undefined),
+    [step.controls.dataSchema, isInlineConfigurableStep]
+  );
+
+  const defaultValues = useMemo(
+    () => ({
+      name: step.name,
+      stepId: step.stepId,
+      ...registerInlineControlValues(step),
+    }),
+    [step, registerInlineControlValues]
+  );
 
-  const form = useForm<z.infer<typeof stepSchema>>({
+  const form = useForm({
     defaultValues,
-    resolver: zodResolver(stepSchema),
+    resolver: zodResolver(buildStepSchema(controlsSchema)),
     shouldFocusError: false,
   });
 
   const { onBlur } = useFormAutosave({
-    previousData: step,
+    previousData: defaultValues,
     form,
-    isReadOnly: isCodeCreatedWorkflow,
+    isReadOnly,
     save: (data) => {
-      update(updateStepInWorkflow(workflow, data));
-      updateStepCache(data);
+      // tranform form fields to step update dto
+      const updateStepData: Partial<StepUpdateDto> = {
+        name: data.name,
+        ...(data.controlValues ? { controlValues: data.controlValues } : {}),
+      };
```

Comment:
We always needs to do these mappings because what we get from API and display in form has different keys than what we send in payload and vice-versa:

- we don't want to send controlValues if they're not changed, otherwise controls validation in InApp step would be triggered on name change
- we dont want to send `stepId` to update payload

Reply:
@ChmaraX If we need to change the API, let's figure out what we have to do and make it easier on the frontend.

Reply:
@SokratisVidros This would be my rough plan:

1. Add `controls`, or the same values we get from step endpoint, to the workflow steps
2. Remove the current complex logic around step vs workflow on frontend - have only `workflow` as a single frontend entity/resource - we always want to operate on workflow resource
3. align types properly for get, update and workflow forms
4. align property names a bit, if GET returns `controls.values` then PUT should not be `controlsValues` but thats opinionated, I dont have strong opinion about this, it just removes the mapping

I can open these changes gradually and discuss them with the team so it doesn't disrupt existing code, like a wrangler pattern.

Reply:
We are aligned.